extensions/virtual-threads/runtime/src/main/java/io/quarkus/virtual/threads/VirtualThreadsRecorder.java

```diff
@@ -0,0 +1,166 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.ShutdownContext;
+import io.quarkus.runtime.annotations.Recorder;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+
+@Recorder
+public class VirtualThreadsRecorder {
+
+    private static final Logger logger = Logger.getLogger("io.quarkus.virtual-threads");
+
+    static VirtualThreadsConfig config = new VirtualThreadsConfig();
+
+    private static volatile Executor current;
+    private static final Object lock = new Object();
+
+    public void setupVirtualThreads(VirtualThreadsConfig c, ShutdownContext shutdownContext, LaunchMode launchMode) {
+        config = c;
+        if (launchMode == LaunchMode.DEVELOPMENT) {
+            shutdownContext.addLastShutdownTask(new Runnable() {
+                @Override
+                public void run() {
+                    Executor executor = current;
+                    if (executor instanceof ExecutorService) {
+                        ((ExecutorService) executor).shutdownNow();
+                    }
+                    current = null;
+                }
+            });
+        } else {
+            shutdownContext.addLastShutdownTask(new Runnable() {
+                @Override
+                public void run() {
+                    Executor executor = current;
+                    if (executor instanceof ExecutorService) {
+                        ExecutorService service = (ExecutorService) executor;
+                        service.shutdown();
+
+                        final long timeout = config.shutdownTimeout.toNanos();
+                        final long interval = config.shutdownCheckInterval.orElse(config.shutdownTimeout).toNanos();
+
+                        long start = System.nanoTime();
+                        int loop = 1;
+                        long elapsed = 0;
+                        for (;;) {
+                            // This log can be very useful when debugging problems
+                            logger.debugf("Await termination loop: %s, remaining: %s", loop++, timeout - elapsed);
+                            try {
+                                if (!service.awaitTermination(Math.min(timeout, interval), NANOSECONDS)) {
+                                    elapsed = System.nanoTime() - start;
+                                    if (elapsed >= timeout) {
+                                        service.shutdownNow();
+                                        break;
+                                    }
+                                } else {
+                                    return;
+                                }
+                            } catch (InterruptedException ignored) {
+                            }
```

Comment:
shouldn't you re-interrupt the thread?

Reply:
The next iteration will wait again on the close until termination or shutdown now.